nasa/ow_simulator
# Implemented custom fault support

## Linked Issues:
| EPIC ⚡ | N/A |
| :----------- | :----------- |
| Jira Ticket 🎟️ | [OCEANWATER-1002](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-1002) |
| Github :octocat:  | # |


## Summary of Changes
* User now has a new option under RQT -> Dynamic Reconfigure -> faults. A custom fault can be added to a specific directory, and when activated, will inject the fault using the same logic from before the changes in [OW-993](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-993).
* Basic error catching prevents the program from terminating if the user inputs an invalid file name, either not found in the _ow_power_system/profiles_ directory, and/or improper formatting.
* ~~Similar to [OW-894](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-894)'s original changes, the user is restricted to using either the provided high power draw or their own custom fault, never both at once. Activating one fault prevents the other from being activated for the remainder of that simulation and appropriate warnings are provided should the user attempt this.~~
  * The restriction is no longer implemented, and the user can activate any combination of the provided high power draw as well as their own custom faults. However, only one custom fault at a time can be active.
* Multiple different custom faults can be deactivated and reactivated in the same simulation, as long as the provided file name is valid (though only one custom fault can be active at a time).
* If the custom fault file runs its course (i.e. the end of the file is reached), the fault injection ceases. If the user deactivates and reactivates the custom fault without changing the file name, the fault will be re-loaded and restart at index 0. Otherwise, whenever the custom fault is deactivated and then reactivated with a new file name in the box, the previous fault is unloaded and the new one loaded.
* A blank example .CSV file has been added under _ow_power_system/profiles/example_fault.csv_. This file is provided to show potential users how to format their own custom faults.
  * The file does not necessarily have to use the .csv extension, but its formatting must be identical to the provided example.
* NOTE: Additional testing was done to see if a more user-friendly method of fault selection was available (e.g. using a file selection dialog, a drop-down menu, etc.) but such options do not seem feasible under the Dynamic Reconfigure tab.

## Test
* catkin build
* roslaunch ow europa_terminator_workspace.launch
* In 3 other terminals, run one of the 3 following commands in each terminal to monitor statistics:
```
rostopic echo /power_system_node/state_of_charge
rostopic echo /power_system_node/remaining_useful_life
rostopic echo /power_system_node/battery_temperature
```
* Either use the provided _example_test.csv_ file for the next step, or create your own custom fault .csv file using the same format as seen in _example_test.csv_ and place it in the _ow_power_system/profiles_ directory
  * One example is to make a new CSV where every index has a 'wattage' value of 25, simulating a 25W high power draw. This should have identical results as a 25W high power draw using the non-custom feature
* Navigate to the Dynamic Reconfigure tab -> faults and enter the full name of the fault file to be tested in the _custom_fault_name_ section
* Check the box for _activate_custom_fault_ to activate the custom fault. Monitor battery statistics to ensure the custom fault's behavior is expected.
* ~~Ensure high power draw cannot be applied in the current simulation (since a valid custom power fault has already been applied) by checking the box for _activate_custom_fault_ and observing the console running the workspace for warnings.~~
  * High power draw can now be used simultaneously with custom power faults.
* Confirm the new fault feature can be utilized from the terminal:
  * Modify the information in the file path text box using a console command in a new terminal:
    * rosrun dynamic_reconfigure dynparam set /faults custom_fault_name [file name]
  * rosrun dynamic_reconfigure dynparam set /faults activate_custom_fault True

## Potential Issue(s)
* ~~Error-catching logic does not prevent user from attempting to activate a .CSV file containing formatting unrecognized by the old code.~~
  * Error detection during file loading has been implemented.
* A better system for file designation (rather than manual entry of the desired file path) would be to use an open-file dialog, however I was unable to find any information on whether or not this is possible within the RQT window. It seems possible, given there are buttons to select _.perspective_ files to upload a perspective to the window, but I couldn't find anything regarding adding one of those buttons within the faults tab of Dynamic Reconfigure.
* Updating the user documentation and release tests will be done after the PR is ready to merge.
